### PR TITLE
feat: reduce memory overhead in CUDA Graph

### LINKED ISSUE #267: feature: reduce memory overhead in CG
### Description

Add a new pool cache layout we manage ourselves on top of PyTorch one when CUDA graphs is used.

### Is this request already on the discussion forum?

_No response_

### Motivation

CG capture for each shape a copy of each model tensor input and keep them in memory.
This is ok when it's just token IDs, it is not for generative model with cache.
We want to improve the way we manage this memory.

### Have you tried to implement it?

_No response_

### Self-service

- [X] I would be willing to contribute to this feature myself.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by pommedeterresautee**
fix #267 

With new pool layer:

```
time to warmup: 12.41min
Kernl speedup: 2.5X (0.5 VS 1.3 min)
# different outputs: 0/73 (0.00%)

memory footprint:
* allocated: 8.6GB
* reserved: 11.3GB
* max reserved: 13.6GB
```

before this PR:

```log
time to warmup: 12.42min
Kernl speedup: 2.4X (0.5 VS 1.3 min)
# different outputs: 0/73 (0.00%)

memory footprint:
* allocated: 10.9GB
* reserved: 13.4GB
* max reserved: 13.9GB
```

test pass (local run 3090 RTX):

```log
======================================================================= 2859 passed, 80 warnings in 342.38s (2:19:02) ========================================================================
```

### CODE REVIEW

File: src/kernl/optimizer/cuda_graph.py [hunk 1]
```diff
@@ -12,48 +12,74 @@
...
-
-        # storage offset should not be used below... otherwise it changes cuda address
-        before_ptr = static_tensor.data_ptr()
-        static_tensor = torch.as_strided(static_tensor, original_tensor.size(), original_tensor.stride())
-        static_tensor.copy_(original_tensor)
-        cuda_graph_input.append(static_tensor)
-        assert before_ptr == static_tensor.data_ptr(), f"should be equal: {before_ptr} != {static_tensor.data_ptr()}"
-
-    return cuda_graph_input
+    inputs_copy = list(inputs).copy()
```
Comment (gaetansnl) [inline src/kernl/optimizer/cuda_graph.py:40]: IMO the function should be split in multiple functions for clarity (for example a function to calculate the size of the pool) and we don't need two loops like this I think. One is enough

File: test/test_cuda_graph.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+def test_aligned_size():
+    t = torch.ones((33,))
+    assert get_aligned_size(t, alignment=16) == 144
+    assert get_aligned_size(t, alignment=32) == 160
+    t = torch.ones((1,))
+    assert get_aligned_size(t, alignment=8) == 8
+
+
+def test_cuda_graph_pool():
```
Comment (gaetansnl) [inline test/test_cuda_graph.py:15]: imo tests are missing: for example errors, edge cases, etc...

File: src/kernl/optimizer/cuda_graph.py [hunk 3]
```diff
@@ -12,48 +12,74 @@
...
+def prepare_inputs(inputs: list[torch.Tensor], pools: list[CudaGraphPool]) -> list[torch.Tensor]:
+    """
+    Copy the inputs in the CUDA graphs memory pool and return tensor copies.
+    Follows a greedy bin packing algorithm (first-fit decreasing) to minimize the number of pools:
+    - sort the items in decreasing order of size ;
+    - insert them one by one into the first bin that has room for it.
+
+    :param inputs: list of tensors to copy in the pool
+    :param pools: list of available pools
+    :return: list of tensors that are copies of the inputs in the pool
```
Comment (gaetansnl) [inline src/kernl/optimizer/cuda_graph.py:38]: "copy of input tensors with the underlying storage in the memory pool" or something like this maybe, not sure. I think the comment is ambiguous

**Review by gaetansnl — CHANGES_REQUESTED** [review]

File: src/kernl/optimizer/cuda_graph.py [hunk 4]
```diff
@@ -12,48 +12,74 @@
...
+                new_t = pool.copy_to_pool(t)
+                new_t.position = t.position
+                outputs.append(new_t)
+                break
+        if new_pool:
+            to_add_new_pool.append(t)
+
+    if len(to_add_new_pool) > 0:
+        total_input_size = sum([get_aligned_size(t) for t in to_add_new_pool])
+        if len(static_inputs_pool) > 0:
```
Comment (gaetansnl) [inline src/kernl/optimizer/cuda_graph.py:68]: imo a comment could be added to explain why this strategy

File: src/kernl/optimizer/pool_cuda_graphs.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        return self.size - self.offset >= get_aligned_size(t)
+
+    def reset(self):
+        """
+        Reset the pool offset to 0.
+        """
+        self.offset = 0
+
+
+def get_aligned_size(t: torch.Tensor, alignment=64) -> int:
```
[the remark this answers is not in the document]
Comment (pommedeterresautee) [reply src/kernl/optimizer/pool_cuda_graphs.py:56]: We don't know what kind of inputs we get. 
It may be int64 (possible)

[the remark this answers is not in the document]
Comment (gaetansnl) [reply src/kernl/optimizer/pool_cuda_graphs.py:56]: I was expecting aligning on 64 bits but not 64 bytes

File: src/kernl/optimizer/pool_cuda_graphs.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
+import torch
+
+
+class CudaGraphPool:
+    """
+    Memory pool for CUDA graphs.
+    """
+
+    def __init__(self, size, dtype=torch.int8, device="cuda"):  # TODO change device to cuda
```
[the remark this answers is not in the document]
Comment (pommedeterresautee) [reply src/kernl/optimizer/pool_cuda_graphs.py:9]: do you mean we should just force int8 to simplify everything?

File: src/kernl/optimizer/pool_cuda_graphs.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+    def copy_to_pool(self, t: torch.Tensor) -> torch.Tensor:
+        """
+        Copy the tensor t in the pool and return a tensor that is a view of the pool.
+        :param t: tensor to copy in the pool
+        :return: tensor copy (that is a view of the pool)
+        """
+        assert t.device == self.pool.device
+        # 64 bits alignment
+        tensor_aligned_size = get_aligned_size(t)
```
[the remark this answers is not in the document]
Comment (pommedeterresautee) [reply src/kernl/optimizer/pool_cuda_graphs.py:29]: you mean get_aligned_size is computed 2 times? is it an issue?

File: src/kernl/optimizer/pool_cuda_graphs.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (pommedeterresautee) [reply src/kernl/optimizer/pool_cuda_graphs.py:9]: removed

File: src/kernl/optimizer/pool_cuda_graphs.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    """
+    Memory pool for CUDA graphs.
+    """
+
+    def __init__(self, size, dtype=torch.int8, device="cuda"):  # TODO change device to cuda
+        """
+        :param size: size of the pool in bytes. Must be a multiple of the dtype size.
+        :param dtype: dtype of the pool
+        """
+        assert size > 8, "Size must be at least 8 bytes"
```
[the remark this answers is not in the document]
Comment (pommedeterresautee) [reply src/kernl/optimizer/pool_cuda_graphs.py:14]: replaced

File: src/kernl/optimizer/pool_cuda_graphs.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (pommedeterresautee) [reply src/kernl/optimizer/pool_cuda_graphs.py:56]: mmmhhh true, it was my intention but messed up. replaced by 8

File: src/kernl/optimizer/cuda_graph.py [hunk 9]
```diff
@@ -12,48 +12,74 @@
...
-        static_tensor = torch.as_strided(static_tensor, original_tensor.size(), original_tensor.stride())
-        static_tensor.copy_(original_tensor)
-        cuda_graph_input.append(static_tensor)
-        assert before_ptr == static_tensor.data_ptr(), f"should be equal: {before_ptr} != {static_tensor.data_ptr()}"
-
-    return cuda_graph_input
+    inputs_copy = list(inputs).copy()
+    # add position meta
+    for index, t in enumerate(inputs_copy):
+        t.position = index
```
[the remark this answers is not in the document]
Comment (pommedeterresautee) [reply src/kernl/optimizer/cuda_graph.py:43]: I don't think so, later the (new) order is lost because inputs are split between those that have been copied in an existing pool and the other to be associated to a pool to create. Do you see a way to avoid that?

File: src/kernl/optimizer/cuda_graph.py (hunk 3 above)
Comment (pommedeterresautee) [reply src/kernl/optimizer/cuda_graph.py:38]: updated

File: src/kernl/optimizer/cuda_graph.py (hunk 4 above)
Comment (pommedeterresautee) [reply src/kernl/optimizer/cuda_graph.py:68]: added

File: src/kernl/optimizer/cuda_graph.py (hunk 1 above)
Comment (pommedeterresautee) [reply src/kernl/optimizer/cuda_graph.py:40]: there are 4 loops, which one should be removed in your opinion?

File: src/kernl/optimizer/pool_cuda_graphs.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (gaetansnl) [reply src/kernl/optimizer/pool_cuda_graphs.py]: yes

**Review by gaetansnl — APPROVED** [review]

**Comment by pommedeterresautee** [comment]
test pass

```log
=============================== warnings summary ===============================
test/test_model_optimization.py: 1 warning
test/test_torchdynamo.py: 79 warnings
  /home/geantvert/.local/share/virtualenvs/kernl/lib/python3.9/site-packages/torch/cuda/graphs.py:82: UserWarning: The CUDA Graph is empty. This ususally means that the graph was attempted to be captured on wrong device or stream. (Triggered internally at ../aten/src/ATen/cuda/CUDAGraph.cpp:191.)
    super(CUDAGraph, self).capture_end()

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================ 2863 passed, 80 warnings in 8556.02s (2:22:36) ================
```
